File: lib/NativeGL/input.rb
require 'ffi'

module CWrapper
	extend FFI::Library
	ffi_lib 'libwin.so'
	
	attach_function :bindEventCallback, [:pointer], :void
	attach_function :pumpEvents, [], :void
end

module RubyWrapper
	def self.bindCallback(&callback)
		CWrapper.bindEventCallback(FFI::Function.new(:void, [], callback))
	end
end

File: lib/NativeGL/src/Input.c
#include "Input.h"

int sdlEventFilter(void* userData, SDL_Event* event)
{
	((void (*)())userData)();
}

void bindEventCallback(void* callback)
{
	SDL_AddEventWatch(sdlEventFilter, callback);
}

void pumpEvents(void)
{
	SDL_PumpEvents();
}